### ISSUE TITLE
Optimize model for inference speed

### ISSUE THREAD

**Issue by EuphoriaCelestial** [NVIDIA/waveglow#194]
https://github.com/NVIDIA/waveglow/issues/54
In this issue, they were talking about lower some parameters to maximize inference speed. But I dont know how to do it properly, what can be reduced and what need to remain. Anyone did this before? Please send me your hparams configuration.

if I trained my model using fp32, can it run inference in fp16 and vice versa? in this case, will it impove inference speed? I am using RTX 2080ti, my model run 7 times faster than real-time, and I am pretty sure it can be improved

and one more thing, is there any benefit of running inference using multi-GPUs?

**Comment by CookiePPP** [comment]
>if I trained my model using fp32, can it run inference in fp16 and vice versa? in this case, will it impove inference speed?

Both fp32 and fp16 training saves the model weights in fp32 format.
fp16 training and inference just means the calculations are done in fp16 and training uses 'loss scaling' to provide higher precision/stability.

Pytorch will automatically use Tensor cores if the model and data is in fp16.
Also
https://github.com/NVIDIA/tacotron2/blob/master/inference.ipynb
```
waveglow_path = 'waveglow_256channels.pt'
waveglow = torch.load(waveglow_path)['model']
waveglow.cuda().eval().half()
```
If you are following the Tacotron2 example notebook then you should already be using Tensor Cores for most of the operations.

**Comment by EuphoriaCelestial** [comment]
> Both fp32 and fp16 training saves the model weights in fp32 format.
> fp16 training and inference just means the calculations are done in fp16 and training uses 'loss scaling' to provide higher precision/stability.

is there any benefit of using fp32 in training? audio quality improve?

**Comment by CookiePPP** [comment]
@EuphoriaCelestial 
https://github.com/NVIDIA/DeepLearningExamples/tree/master/PyTorch/SpeechSynthesis/Tacotron2#results
Not sure :man_shrugging:

**Comment by EuphoriaCelestial** [comment]
> @EuphoriaCelestial
> https://github.com/NVIDIA/DeepLearningExamples/tree/master/PyTorch/SpeechSynthesis/Tacotron2#results
> Not sure man_shrugging
> fp32 seems to learn more per epoch but I'm not sure if they use different batch size for fp16 in that example.

I used fp32 for my model and have to reduce batch size. For single RTX 2080ti, it can run with batch size = 14. Is it too little? will it affect training result?

**Comment by CookiePPP** [comment]
@EuphoriaCelestial 
https://arxiv.org/abs/1711.00489
You can decrease learning rate with smaller batch size and have similar result to high learning rate and high batch size.

**Comment by EuphoriaCelestial** [comment]
> @EuphoriaCelestial
> https://arxiv.org/abs/1711.00489
> You can decrease learning rate with smaller batch size and have similar result to high learning rate and high batch size.

oopsie ... I have already started training, is it okay to adjust learning rate now?
